awx/ui/client/src/instance-groups/instance-groups.service.js

```diff
@@ -9,33 +9,59 @@ export default
+                        // position, mark it and all remaining currentIds in the array for disassociation.
+                        let disassociateRemainingIds = false;
+                        currentGroupIds.forEach((currentId, position) => {
+                            if (!disassociateRemainingIds && updatedGroupIds[position] !== currentId) {
+                                disassociateRemainingIds = true;
+                            }
 
-                    let groupsToDisassociate = current
-                        .filter(val => groupsToAdd
-                            .indexOf(val) === -1)
-                        .map(val => ({id: val, disassociate: true}));
+                            if (disassociateRemainingIds) {
+                                groupIdsToDisassociate.push(currentId);
+                            }
+                        });
 
-                    let groupsToAssociate = groupsToAdd
-                        .filter(val => current
-                            .indexOf(val) === -1)
-                        .map(val => ({id: val, associate: true}));
+                        updatedGroupIds.forEach(updatedId => {
+                            if (groupIdsToDisassociate.includes(updatedId)) {
+                                // we get here if the id was marked for disassociation due to being
+                                // out of order - we'll need to re-associate it.
+                                groupIdsToAssociate.push(updatedId);
+                            } else if (!currentGroupIds.includes(updatedId)) {
+                                // we get here if the id is a new association
+                                groupIdsToAssociate.push(updatedId);
+                            }
+                        });
 
-                    let pass = groupsToDisassociate
-                        .concat(groupsToAssociate);
+                        // convert the id arrays into request data
+                        const groupsToAssociate = groupIdsToAssociate.map(id => ({ id, associate: true}));
+                        const groupsToDisassociate = groupIdsToDisassociate.map(id => ({ id, disassociate: true }));
 
-                    Rest.setUrl(url);
-                    let defers = pass.map((group) => Rest.post(group));
-                    Promise.resolve(defers);
-                });
-            }
-        };
-}];
\ No newline at end of file
+                        // make the disassociate-then-associate request sequence
+                        Rest.setUrl(url);
+                        const disassociationPromise = Promise.all(groupsToDisassociate.map(data => Rest.post(data)));
+                        
+                        return disassociationPromise
+                            .then(() => {
+                                let associationPromise = $q.resolve();
+                                groupsToAssociate.forEach(data => {
```

Comment:
we need to do the association requests sequentially to make sure they're processed in the right order so we build a promise chain instead of using `.all`.